### DISCUSSION TITLE
OPDs calculated by rayoptics do not match the values calculated by Zemax OpticStudio

### DISCUSSION THREAD

**Discussion by Nuvolona** [mjhoptics/ray-optics#110] [Q&A]
Hello everyone,

I am having a problem calculating the OPDs using rayoptics. I do get values using the methods RayFanFigure.eval_fct, analyses.eval_fan and trace_with_opd. These give the same results, however these differ from the values calculated by Zemax OpticStudio.

Below is the code I use to read a .zmx file (using rayoptics's zmxread) and to evaluate the OPDs.

```python
import rayoptics.environment as roe
import rayoptics.zemax as roz
import pathlib
import numpy as np
import matplotlib.pyplot as plt

opm, info = roz.zmxread.read_lens_file(pathlib.Path(__file__).parent
                                       / 'start_system_3.zmx')
sm = opm.seq_model

print(sm.list_model())

wav_plt = plt.figure(FigureClass=roe.RayFanFigure,
                     opt_model=opm, data_type='OPD', scale_type=roe.Fit.All_Same).plot()
                     
layout_plt = plt.figure(FigureClass=roe.InteractiveLayout, opt_model=opm).plot()

print('Using RayFanFigure')
for i in range(4):
    fld, wvl, foc = opm.optical_spec.lookup_fld_wvl_focus(i)
    rff = roe.RayFanFigure(opm, 'OPD', num_rays=11)
    tan = rff.eval_fct(i, 1)[1][0]
    sag = rff.eval_fct(i, 0)[1][0]

    print('FLD = {0}\n{1:>10} {2:>10}'.format(fld, 'tan', 'sag'))
    for pair in zip(tan, sag):
        print("{0:>10.7f} {1:>10.7f}".format(*pair))
    print()

print('Using eval_fan')
for i in range(4):
    fld, wvl, foc = opm.optical_spec.lookup_fld_wvl_focus(i)
    tan_res = roe.analyses.eval_fan(opm, fld, wvl, foc, 1, num_rays=11)
    sag_res = roe.analyses.eval_fan(opm, fld, wvl, foc, 0, num_rays=11)
    tan, sag = [], []
    for val in tan_res:
        tan.append(val[1][2])
    for val in sag_res:
        sag.append(val[1][2])

    print('FLD = {0}\n{1:>10} {2:>10}'.format(fld, 'tan', 'sag'))
    for pair in zip(tan, sag):
        print("{0:>10.7f} {1:>10.7f}".format(*pair))
    print()

print('Using trace_with_opd')
for i in range(4):
    fld, wvl, foc = opm.optical_spec.lookup_fld_wvl_focus(i)
    # defocus = opm.optical_spec.defocus.defocus_range
    print('FLD = {0}\n{1:>10}'.format(fld, 'tan'))
    for pupil in zip(np.zeros(11),
                     np.linspace(-1, 1, num=11)):
        ray, ray_opl, wvl, ray_opd = roe.trace_with_opd(opm, pupil,
                                                        fld, wvl, foc)
        ray_opd = ray_opd / opm.nm_to_sys_units(wvl)
        print("{0:>10.7f}".format(ray_opd))
    print()

plt.show()
```

The values I get are for example (FLD=10.0 deg):
|tan|sag|
|---|---|
|0.0311662 | 0.0464790|
|-0.0377442 | 0.0166939|
|-0.0447002 |-0.0067009|
|-0.0250897 |-0.0103304|
|-0.0062340 |-0.0038950|
| 0.0000000 |-0.0000000|
|-0.0030835| -0.0038950|
|-0.0041825 |-0.0103304|
| 0.0007489 |-0.0067009|
|-0.0183967|  0.0166939|
|-0.1717649|  0.0464790|

The values calculated by OpticStudio however are:
|tan|sag|
|---|---|
|0.062550| 0.050512|
|-0.022892| 0.016965|
|-0.040332 |-0.007487|
|-0.025071 |-0.010963|
|-0.006684 |-0.004094|
|0.000000 |0.000000|
|-0.004122|  -0.004094|
|-0.008328 |-0.010963|
|-0.005686 |-0.007487|
|-0.017458 |0.016965|
|-0.134766| 0.050512|

Plotting the layout both in rayoptics and in Zemax results in:
![Rayoptics Layout](https://user-images.githubusercontent.com/120388502/213036370-33865082-d2ed-4836-9e9b-050e6f0c0ece.png)
![Zemax Layout](https://user-images.githubusercontent.com/120388502/213036378-65196fda-e477-407f-97ed-bc7f9d78d56a.png)
The third lens in the rayoptics figure is somehow too short, and I dont know why. I am not sure, if this behavior is the reason for the difference in the values. I tried calculating the OPD for a single lens, but they also did not match the Zemax values. Is there any problem using zmxread, or am I doing something wrong? Am I missing some arguments in the methods used? 

I have attached the .zmx file and python code in a .zip file, the opd values from zemax, as well as the single lens .zmx file, values and plot in zip form. Any kind of help is appreciated.
[start_system_3_zmx.zip](https://github.com/mjhoptics/ray-optics/files/10440329/start_system_3_zmx.zip)
[Zemax OPD Values.txt](https://github.com/mjhoptics/ray-optics/files/10440332/Zemax.OPD.Values.txt)
[start_system_1_zmx.zip](https://github.com/mjhoptics/ray-optics/files/10440558/start_system_1_zmx.zip)

**Comment by mjhoptics** [comment]
Hello @Nuvolona

I believe some of the differences between programs is that there are differences in how vignetting is modeled. The Zemax lens layout shows off-axis beams at the stop surface that are outside the axial beam footprint. This isn't really the right way to model a stop surface - the stop is by definition the limiting aperture in the system. 

The axial OPDs agree to 1e-6 waves. The off-axis fields show differences that could arise from several sources. To help diagnose the differences, it would be useful to have Zemax data for surface by surface ray coords and direction cosines for the chief ray and upper and lower yfan rays for the axis and full fields. These details can be compared with the corresponding **ray-optics** data and the **ray-optics** vignetting adjusted to match the Zemax-based model.

Another possible factor for OPD calculations are choice of the center of the reference sphere. Lateral shifts of the reference point manifest as wavefront tilt. If the system has distortion, the chief ray and the paraxial image heights will differ. ray-optics uses the chief ray as the reference, so if Zemax uses the paraxial/Gaussian height, that could introduce wavefront tilt.

I am working on a Jupyter notebook based on your script. Hopefully it will help understand the differences.

Regards

Mike

**Comment by mjhoptics** [comment]
Hello @Nuvolona

Here is a notebook with some ideas on how to compare results between 2 programs. If you can generate the same kind of surface-by-surface ray trace data from Zemax as in the notebook, we can take the comparison to the next phase. Let me know if you have any questions.
Regards
Mike

[Results_Comparison.pdf](https://github.com/mjhoptics/ray-optics/files/10494951/Results_Comparison.pdf)
[Results_Comparison_ipynb.zip](https://github.com/mjhoptics/ray-optics/files/10494964/Results_Comparison_ipynb.zip)

**Comment by Nuvolona** [comment]
Hello @mjhoptics,

thank you very much for your reply. I have been trying to generate similar surface-by-surface ray trace data from Zemax with no matching results unfortunately. The Zemax layout does actually not show an off-axis beam at the stop surface (Clear Semi-Diameter Margin of 50%):

![zemax semi diameter margin](https://user-images.githubusercontent.com/120388502/217771478-6d2b6489-eaf3-470b-a634-7d20a809bb0e.png)

The stop in Zemax, if I am not wrong, is the limiting aperture in the system then. Zemax has a method to automatically set a vignette, however there is none. The raytrace values of zemax for the last field are:

![Raytrace 1](https://user-images.githubusercontent.com/120388502/217773191-fc3fb6c6-1777-4c6f-ad67-73940534d1ef.png)
![Raytrace 0](https://user-images.githubusercontent.com/120388502/217773182-2f0d121f-68e5-47c4-a814-acc9d9b8231c.png)
![raytrace -1](https://user-images.githubusercontent.com/120388502/217773198-9e5433f8-703e-4f74-ac4c-ec903776f53b.png)

These values do not match the data calculated by rayoptics with and without vignette. Setting RayAiming in Zemax to paraxial results in
![Raytrace 0 rayaiming paraxial](https://user-images.githubusercontent.com/120388502/217774616-a384a411-efe9-46a5-9d95-795f621ebb51.png)

The surface-by-surface ray trace data does match but only for a pupil at [0, 0]. The other values still do not match:
![Raytrace 1 rayaiming paraxial](https://user-images.githubusercontent.com/120388502/217776522-a63f18c4-e4e6-4336-8f90-3a3066c2e545.png)
![Raytrace -1 rayaiming paraxial](https://user-images.githubusercontent.com/120388502/217776538-f31576d9-4833-48f9-ad49-6d110237b39b.png)

The System's properties in Zemax and rayoptics match, so there should be no problem. There aren't many other options in Zemax to change, to get matching results without them being read by rayoptics.

**Comment by mjhoptics** [comment]
Hello @Nuvolona ,
I've looked at the data you sent (thank you!) and I've attached the results of the investigation so far. 
You included 2 sets of data. The first one appears to be using the paraxial entrance pupil as the aiming surface. The full field chief ray doesn't go through the center of the stop surface (surface 3) in this situation. The second data set shows the chief ray going through the stop center, matching, as you note, the ray-optics result. 
The difference I found is that the Zemax model is using the paraxial value for the beam size at the stop rather than the value from the real ray trace (which is what ray-optics uses). I used the ray aiming function directly to find the rays that go through the stop at the paraxial stop value and included that data in the attached pdf (in the zip file). These rays match the surface by surface trace that you included from Zemax. I calculated the OPD for these rays and included the output; hopefully you can compare with the Zemax results to see how close they are with this change.
I also included a modified lens file with these changes; I think it should be readable by the current release version of ray-optics.
I've found a few problems with how apertures are handled when importing as well as adding some improvements when iterating rays that aren't in the released version. I've been updating them to the Develop branch and hopefully will be able to build a release in the not too distant future.
Hope this helps,
Mike

[Results_Comparison_zmx_rt_enp.zip](https://github.com/mjhoptics/ray-optics/files/10725943/Results_Comparison_zmx_rt_enp.zip)

**Comment by Nuvolona** [comment]
Hello @mjhoptics,

sorry for the late reply. Thank you for your answer, I have been able to reproduce your results using the delevop branch. The good news are that the ray trace data match and that the opds match to the fourh decimal point as well. So these changes worked for this system.  However using another Zemax lens system, the opds unfortunately still do not match. The ray trace data however does match, except for the last surface. 

The other lens system is shown below.
![image](https://user-images.githubusercontent.com/120388502/230020028-42f72668-d83f-4f45-a54f-b8162190dd92.png)

I have attached the new lens system, its ray trace data and some opd values.
[new_lens_system.zip](https://github.com/mjhoptics/ray-optics/files/11156750/new_lens_system.zip)

**Reply by mjhoptics** [reply]
Hello @Nuvolona,

I looked at your model and have a couple of comments.

After importing the zemax file, I had to do a set_vignetting(opm) to get the boundary rays to match.

Following this, the OPD analysis in ray-optics shows significant differences from the zemax data. There is a larger problem though and that is the image plane is very far from the geometrical image point, indeed almost an entire focal length away. There is a implied assumption in many OPD calculations that the reference point for the OPD is "close to" the wavefront center of curvature.

For the purposes of comparing programs under common conditions, I would change the image gap thickness from 100. to 33.62689, which is the distance to the paraxial image point. Comparing the OPDs wrt the paraxial image plane is a common use case and all programs should agree pretty well under these conditions, I would think.

I've included a Jupyter notebook and a pdf rendering of it in the attached zip file. 
[Results_Comparison_zmx_new_sys.zip](https://github.com/mjhoptics/ray-optics/files/11166948/Results_Comparison_zmx_new_sys.zip)
